presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java

```diff
@@ -1160,6 +1179,108 @@ public void append(JsonParser parser, BlockBuilder blockBuilder)
         }
     }
 
+    private static class RowBlockBuilderAppender
+            implements BlockBuilderAppender
+    {
+        BlockBuilderAppender[] fieldAppenders;
+        Optional<Map<String, Integer>> fieldNameToIndex;
+
+        RowBlockBuilderAppender(BlockBuilderAppender[] fieldAppenders, Optional<Map<String, Integer>> fieldNameToIndex)
+        {
+            this.fieldAppenders = fieldAppenders;
+            this.fieldNameToIndex = fieldNameToIndex;
+        }
+
+        @Override
+        public void append(JsonParser parser, BlockBuilder blockBuilder)
+                throws IOException
+        {
+            if (parser.getCurrentToken() == JsonToken.VALUE_NULL) {
+                blockBuilder.appendNull();
+                return;
+            }
+
+            if (parser.getCurrentToken() != START_ARRAY && parser.getCurrentToken() != START_OBJECT) {
+                throw new JsonCastException(format("Expected a json array or object, but got %s", parser.getText()));
+            }
+
+            parseJsonToSingleRowBlock(
+                    parser,
+                    (SingleRowBlockWriter) blockBuilder.beginBlockEntry(),
+                    fieldAppenders,
+                    fieldNameToIndex);
+            blockBuilder.closeEntry();
+        }
+    }
+
+    public static Optional<Map<String, Integer>> getFieldNameToIndex(List<RowField> rowFields)
+    {
+        if (!rowFields.get(0).getName().isPresent()) {
+            return Optional.empty();
+        }
+
+        Map<String, Integer> fieldNameToIndex = new HashMap<>(rowFields.size());
+        for (int i = 0; i < rowFields.size(); i++) {
+            fieldNameToIndex.put(rowFields.get(i).getName().get(), i);
+        }
+        return Optional.of(fieldNameToIndex);
+    }
+
+    // TODO: inline this method once CAST function support to take a BlockBuilder as input or it supports cachedInstanceFactory.
```

Comment:
The reader of this will likely be confused. Please be a little more explicit.

cacheInstanceFactory or BlockBuilder -> unified RowBlockBuilderAppender and JsonToRowCast -> single caller -> inline